Imperial Government conversion, bugfixes, and Logging improvements

DrDudelsack opened the pull request:
Countries in I:R which have the imperial government form shall be converted with having CK II's imperial government form, while e_roman_empire and e_byzantium will no longer always have the imperial government form. Converted rulers of countries with the imperial government form shall have their children, grandchildren, great-grandchildren, ect. start with the Born in the Purple trait, giving them preference in the succession. 

The bug randomly making countries and subjects republics/merchant republics has been fixed, and the information that is logged has been significantly reduced, creating readable error logs

DrDudelsack committed:
Imperial Government

Countries in I:R which have the imperial government form shall be converted with having CK II's imperial government

DrDudelsack committed:
checkFile Method

Created a method to check of a file exists or not. It turns out that it wasn't needed for Born in the Purple calculation, but it could be useful to have anyway

DrDudelsack committed:
Imperial Government

DrDudelsack committed:
Imperial Government

Added directories needed for imperial government creation

DrDudelsack committed:
Create test

DrDudelsack committed:
Add files via upload

File used to allow for imperial governments

DrDudelsack committed:
Create test

DrDudelsack committed:
Add files via upload

File used to allow for imperial governments

DrDudelsack committed:
Create test

DrDudelsack committed:
Add files via upload

File used to allow for imperial governments

Idhrendur commented on ImperatorToCk2/DataFiles/defaultOutput/government_flavor/test:
```diff
@@ -0,0 +1 @@
+
```
Do you still need this file now that the PR has been put together? If not, delete it.

Idhrendur commented on ImperatorToCk2/DataFiles/defaultOutput/governments/test:
```diff
@@ -0,0 +1 @@
+
```
Likewise, if you no longer need this file delete it.

Idhrendur commented on ImperatorToCk2/DataFiles/defaultOutput/laws/test:
```diff
@@ -0,0 +1 @@
+
```
Same, delete if not needed.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -331,7 +331,7 @@ public static void main (String[] args) throws IOException
                         String[] owners = irOwners[aq5].split(","); 
 
-                        LOGGER.info(irOwners[aq5]+"_irOwners_"+aq2);  
+                        //LOGGER.info(irOwners[aq5]+"_irOwners_"+aq2);  
```
If you no longer need this line, delete it entirely.

DrDudelsack replied:
I still could use these commented logging statements in case somebody has a save that causes problems/crashes. On my build, I can re-enable these logging statements when I parse the save through to get a thorough analysis on the problem. 

After diagnosing and fixing the issue, they could then be disabled again so they don't create 1,000+ line error logs for general users

Idhrendur replied:
There are better ways to accomplish what you want, but fair enough for now.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -351,13 +351,13 @@ public static void main (String[] args) throws IOException
 
                         ownerTot[ownNum] = Integer.parseInt(owners[1]);
-                        LOGGER.info(owners[0]+owners[1]+"b_owners");    
+                        //LOGGER.info(owners[0]+owners[1]+"b_owners");    
```
This one too.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -351,13 +351,13 @@ public static void main (String[] args) throws IOException
                             if (ownerTot[aq6] > ownerTot[aq6-1]){
                                 ck2ProvInfo[0][aq2] = owners[0];
-                                LOGGER.info((ck2ProvInfo[0][aq2])+"_"+aq2+"cq");
+                                //LOGGER.info((ck2ProvInfo[0][aq2])+"_"+aq2+"cq");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -375,7 +375,7 @@ public static void main (String[] args) throws IOException
                 }
                 else if (aq2 < 380) {
-                    LOGGER.info (ck2TagTotals[aq2] + "_" + aq2);    
+                    //LOGGER.info (ck2TagTotals[aq2] + "_" + aq2);    
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -411,7 +411,7 @@ else if (aq2 < 380) {
             LOGGER.info("Importing country data...");
 
-            LOGGER.config("The region is" + ck2ProvInfo[4][343]);
+            //LOGGER.config("The region is" + ck2ProvInfo[4][343]);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -535,15 +540,15 @@ else if (aq2 < 380) {
 
                                 int subjectOrNot = Processing.checkSubjectList(aq4,impSubjectInfo);
-                                LOGGER.config("subjectOrNot at " + aq4 + " is " + subjectOrNot);
+                                //LOGGER.config("subjectOrNot at " + aq4 + " is " + subjectOrNot);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -535,15 +540,15 @@ else if (aq2 < 380) {
-                                    LOGGER.info("Free Nation at " + aq4);
+                                    Output.titleCreation(impTagInfo.get(aq4)[0],tempNum2,impTagInfo.get(aq4)[3],impTagInfo.get(aq4)[17],
+                                          impTagInfo.get(aq4)[5],rank,"no_liege",modDirectory);
+                                    //LOGGER.info("Free Nation at " + aq4);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -559,26 +564,26 @@ else if (aq2 < 380) {
+                                        Output.titleCreation(impTagInfo.get(aq4)[0],tempNum2,impTagInfo.get(aq4)[3],impTagInfo.get(aq4)[17],
                                             impTagInfo.get(aq4)[5],rank,overlord,modDirectory);
-                                        LOGGER.info("Subject Nation at " + aq4 + " Overlord is " + subjectInfo[0]);
+                                        //LOGGER.info("Subject Nation at " + aq4 + " Overlord is " + subjectInfo[0]);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -559,26 +564,26 @@ else if (aq2 < 380) {
+                                        Output.titleCreation(impTagInfo.get(aq4)[0],tempNum2,impTagInfo.get(aq4)[3],impTagInfo.get(aq4)[17],
                                             impTagInfo.get(aq4)[5],rank,overlord,modDirectory);
-                                        LOGGER.info("Tributary Nation at " + aq4 + " Overlord is " + subjectInfo[0]);
+                                        //LOGGER.info("Tributary Nation at " + aq4 + " Overlord is " + subjectInfo[0]);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -559,26 +564,26 @@ else if (aq2 < 380) {
                                 }
 
-                                LOGGER.info (impTagInfo.get(aq4)[16] + " rules " + impTagInfo.get(aq4)[0] + "_" + aq4);
+                                //LOGGER.info (impTagInfo.get(aq4)[16] + " rules " + impTagInfo.get(aq4)[0] + "_" + aq4);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -559,26 +564,26 @@ else if (aq2 < 380) {
-                                    Character[15],saveCharacters,"q","q",convertedCharacters,impCharInfoList,modDirectory);
-                                LOGGER.config ("c");
+                                    Character[15],impTagInfo.get(aq4)[17],"q","q",convertedCharacters,impCharInfoList,modDirectory);
+                                //LOGGER.config ("c");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -588,16 +593,16 @@ else if (aq2 < 380) {
-                                LOGGER.info(tempTest+impTagInfo.get(aq4)[16] + "_" +Character[3]+Character[0]+Character[7]);
-                                LOGGER.info ("Name is " + locName[0] + " for " +impTagInfo.get(aq4)[0]);
-                                LOGGER.config ("output1");
+                                //LOGGER.info(tempTest+impTagInfo.get(aq4)[16] + "_" +Character[3]+Character[0]+Character[7]);
+                                //LOGGER.info ("Name is " + locName[0] + " for " +impTagInfo.get(aq4)[0]);
+                                //LOGGER.config ("output1");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -639,11 +644,11 @@ else if (aq2 < 380) {
                 }catch (java.util.NoSuchElementException exception){
                     flag = 1;
-                    LOGGER.config("NoSuchElementException and flag = 1");
+                    //LOGGER.config("NoSuchElementException and flag = 1");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -639,11 +644,11 @@ else if (aq2 < 380) {
             }catch (java.lang.ArrayIndexOutOfBoundsException exception){
                 flag = 1;
-                LOGGER.config("ArrayIndexOutOfBoundsException and flag = 1" + "_" + aq4);
+                //LOGGER.config("ArrayIndexOutOfBoundsException and flag = 1" + "_" + aq4);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -739,7 +744,7 @@ else if (aq2 < 380) {
 
                             }
-                            LOGGER.info("Creating province "+importedInfo[0]+" at ID "+aq4+" ruled by "+ruler);
+                            //LOGGER.info("Creating province "+importedInfo[0]+" at ID "+aq4+" ruled by "+ruler);
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -752,24 +757,24 @@ else if (aq2 < 380) {
                 }catch (java.util.NoSuchElementException exception){
                     flag = 1;
-                    LOGGER.config ("Exception1");
+                    //LOGGER.config ("Exception1");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -752,24 +757,24 @@ else if (aq2 < 380) {
-                LOGGER.config(ck2ProvInfo[1][341] + "_341");
-                LOGGER.config(ck2ProvInfo[1][340] + "_340");
-                LOGGER.config(ck2ProvInfo[1][339] + "_339");
+                //LOGGER.config ("Exception2");
+                //LOGGER.config(ck2ProvInfo[1][343] + "_343");
+                //LOGGER.config(ck2ProvInfo[1][342] + "_342");
+                //LOGGER.config(ck2ProvInfo[1][341] + "_341");
+                //LOGGER.config(ck2ProvInfo[1][340] + "_340");
+                //LOGGER.config(ck2ProvInfo[1][339] + "_339");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Main.java:
```diff
@@ -752,24 +757,24 @@ else if (aq2 < 380) {
-            LOGGER.config(ck2ProvInfo[1][340] + "_340");
-            LOGGER.config(ck2ProvInfo[1][339] + "_339");
-            LOGGER.config(ck2ProvInfo[1][338] + "_338");
+            //LOGGER.config(ck2ProvInfo[1][343] + "_343");
+            //LOGGER.config(ck2ProvInfo[1][342] + "_342");
+            //LOGGER.config(ck2ProvInfo[1][341] + "_341");
+            //LOGGER.config(ck2ProvInfo[1][340] + "_340");
+            //LOGGER.config(ck2ProvInfo[1][339] + "_339");
+            //LOGGER.config(ck2ProvInfo[1][338] + "_338");
```
Ditto

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Output.java:
```diff
@@ -939,4 +963,102 @@ else if (tag.equals("9999")) { //uncolonized province, creates dejure provinces
+
+        try {
+
+            while (flag == 0) {
```
This is bad flow control. Replace with a proper loop. I think you can just do
```
for (String line: oldFile) {
  ...
}
```

That may not be exactly correct, so adjust the syntax correctly. You'll no longer need to do ```oldfile.get()```, and can instead just use ```line``` directly

DrDudelsack replied:
How's the try/catch block bad flow control? It seems to work fine from the tests I've done

Idhrendur replied:
Responded on the discord

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Output.java:
```diff
@@ -939,4 +963,102 @@ else if (tag.equals("9999")) { //uncolonized province, creates dejure provinces
+        PrintWriter out = new PrintWriter(fileOut);
+
+        int flag = 0;
+        int aqq = 0;
```
Delete once you fix the loop below.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Output.java:
```diff
@@ -939,4 +963,102 @@ else if (tag.equals("9999")) { //uncolonized province, creates dejure provinces
+                    out.println (imperialTag);
+                }
+                
+                aqq = aqq + 1;
```
Delete once you fix the loop.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Output.java:
```diff
@@ -939,4 +963,102 @@ else if (tag.equals("9999")) { //uncolonized province, creates dejure provinces
+        }catch (java.lang.IndexOutOfBoundsException exception){
+            flag = 1;
+
+        } 
```
Delete once you fix the loop.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Output.java:
```diff
@@ -939,4 +963,102 @@ else if (tag.equals("9999")) { //uncolonized province, creates dejure provinces
+        }catch (java.lang.IndexOutOfBoundsException exception){
+            flag = 1;
+
+        } 
```
Fix this loop similarly to the other one.

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Processing.java:
```diff
@@ -1039,4 +1038,20 @@ else if (provName.equals("kyzyl_su")) {
         return cultureTitles;
     }
 
+    public static String checkFile(String directory) throws IOException //checks if file exists or not
```
Return boolean instead of String.

Idhrendur replied:
The parameter should be called ```file``` not ```directory```

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Processing.java:
```diff
@@ -1039,4 +1038,20 @@ else if (provName.equals("kyzyl_su")) {
+        }
+
+        return "yes";
+    }
```
The entire function is too complex. There's no need for doing something that might throw an exception (exceptions should be for exceptional circumstances, not something normal). Instead, just do

```
File f = new File(directory);
if(f.exists() && !f.isDirectory()) { 
    return true;
}

return false;
```

DrDudelsack committed:
All imperial governments become empires

I've changed it so that independent countries which have the imperial government form in I:R will automatically be empire tier in CK II, even if they don't meet the size requirements. I've done this because CK II imperial titles cannot create or usurp empire titles, so an imperial government at kingdom tier would be effectively locked to that tier for the rest of the 1,000+ year megacampaign

DrDudelsack committed:
All imperial governments become empires

I've changed it so that independent countries which have the imperial government form in I:R will automatically be empire tier in CK II, even if they don't meet the size requirements. I've done this because CK II imperial titles cannot create or usurp empire titles, so an imperial government at kingdom tier would be effectively locked to that tier for the rest of the 1,000+ year megacampaign

DrDudelsack committed:
Removed unused scnr variable

DrDudelsack committed:
Delete test

DrDudelsack committed:
Delete test

DrDudelsack committed:
Delete test

DrDudelsack committed:
Copy flags, formatting changes

DrDudelsack committed:
Copy flags, subjects return overlord's capital, bugfixes

DrDudelsack committed:
Rank Parameter in country list

The save file doesn't contain the info, but it's convenient to have as a field to be modified

DrDudelsack committed:
checkFile improvements

DrDudelsack committed:
Flag directory

DrDudelsack committed:
Manichean support

Added mapping for Manichean religion from TinWiz's timeline extension mod

DrDudelsack committed:
Kamarupa and Orissa mapping

DrDudelsack committed:
New Molon Labe mod formables

Support for new formable nations from the Molon Labe mod

DrDudelsack committed:
Create test

DrDudelsack committed:
Default placefolder flag

Placefolder flag for countries which don't have any other flag assigned to them

DrDudelsack committed:
Delete blank

DrDudelsack committed:
Delete test

DrDudelsack committed:
Dejure blanking files

Blanked vanilla duchies that were causing problems with the 1066 start date dejure calculation

DrDudelsack committed:
c_anglesey typo fix

Idhrendur commented on ImperatorToCk2/src/main/java/ImperatorToCK2/Output.java:
```diff
@@ -939,4 +1041,133 @@ else if (tag.equals("9999")) { //uncolonized province, creates dejure provinces
+        return ck2CultureInfo;
+    }
+
+    public static void copyRaw(String dir1, String dir2) throws IOException
+    {
+
+        String VM = "\\";
+        VM = VM.substring(0);
+
+        FileInputStream fileIn= new FileInputStream(dir1);
+        FileOutputStream fileOut= new FileOutputStream(dir2);
+
+        boolean endOrNot = true;
+        int aqq = 0;
+
+        try {
+            while (endOrNot = true && aqq != -1){
+                if (aqq != -1) {
+                    aqq = fileIn.read();
+                    fileOut.write(aqq);
+                }
+
+            }
+        }catch (java.util.NoSuchElementException exception){
+            endOrNot = false;
+            fileIn.close();
+            fileOut.close();
+
+        }   
+
+    }
```
Instead of having this function, use one of the methods here: https://www.baeldung.com/java-copy-file
I'd recommend #3

DrDudelsack replied:
The function I wrote works fine, I think that it's more trouble then it's worth to change it, especially with the other work that needs to be done before release

Idhrendur replied:
I'm going to be hard-nosed about this. 
1. Overly-complex code tends to infect a project, as it makes it easier for more overly-complex code to slip in.
2. It'll take you less time to do the update than it did to reply. Here's the relevant line: ```Files.copy(dir1, dir2, StandardCopyOption.REPLACE_EXISTING)```

Idhrendur commented on ImperatorToCk2/DataFiles/religionConversion.txt:
```diff
@@ -1,4 +1,4 @@
-roman_pantheon,hellenic_pagan
+\roman_pantheon,hellenic_pagan
```
Do you actually want the backslash here, or was that a mistake?

DrDudelsack replied:
Thank you for catching that, that's indeed a typo, lol

Idhrendur reviewed (changes requested):
The comments from before.

DrDudelsack committed:
Fixed typo

DrDudelsack committed:
Added comagoosie

Added credit to comagoosie for developing the Rakaly save decompressor

DrDudelsack committed:
Rakaly formatting

Added formatting compatibility for saves decompressed using Rakaly

DrDudelsack committed:
Rakaly formatting

Added formatting compatibility for saves decompressed using Rakaly

DrDudelsack committed:
File excecution

Added a method to execute the Rakaly decompressor

DrDudelsack committed:
Rakaly support

DrDudelsack committed:
Rakaly for Windows

Rakaly save decompressor for Windows, created by comagoosie

DrDudelsack committed:
Rakaly support

DrDudelsack committed:
Typo fix

DrDudelsack committed:
Rakaly file

Copies over the rakaly_win.exe file used for Windows save decompression

DrDudelsack committed:
Typo fix
